new tool : "samtools samples"

lindenb opened the pull request:
Hi all,

# short description 

a new subcommand "samples"  listing the samples in a set of bam files.

# Motivation

when using a workflow manager, it is often required to provide an input containing the samples in a bam file.

# Example:

## usage:

```
Usage: samtools samples [options] <input> [...]
       find dir1 dir2 -type f \( -name "*.bam" -o -name "*.cram" \) | samtools samples [options]

Options:
  -h              print help and exit
  -H              print a header
  -i              test if the file is indexed.
  -T <tag>        set the TAG in the @RG line [SM].
  -o <file>       output file [stdout].
  -d              enable multiple samples in one bam
  -m              enable missing sample in one bam. "." will be used as the default name.
  -f <file.fa>    add an indexed fasta in the collection of reference. Can be used multiple times.
  -F <file.txt>   read a file containing the path to the indexed fasta references. One path per line.

 Using -f or -F will add a column containing the path to the reference or "." if the reference was not found.
```

## list the samples

```
$ ./samtools samples -H DIR/S*.bam
#SM	PATH
S1	DIR/S1.bam
S2	DIR/S2.bam
S3	DIR/S3.bam
S4	DIR/S4.bam
S5	DIR/S5.bam
```


## list the libraries (tag=LB)

```
$ ./samtools samples -H DIR/S*.bam -T LB
#LB	PATH
L1	DIR/S1.bam
L2	DIR/S2.bam
L3	DIR/S3.bam
L4	DIR/S4.bam
samtools samples: multiple @RG:"LB" in "DIR/S5.bam". Use option -d to enable multiple.
```

```
$ ./samtools samples DIR/S*.bam -H -T LB -d
#LB	PATH
L1	DIR/S1.bam
L2	DIR/S2.bam
L3	DIR/S3.bam
L4	DIR/S4.bam
L5	DIR/S5.bam
L5b	DIR/S5.bam
```

```
$ ./samtools samples DIR/ENCFF331CGL.rnaseq.b38.bam
#SM	PATH
samtools samples: no @RG:SM in "DIR/ENCFF331CGL.rnaseq.b38.bam". Use option -m to enable missing.

$ ./samtools samples DIR/ENCFF331CGL.rnaseq.b38.bam -m
#SM	PATH
.	DIR/ENCFF331CGL.rnaseq.b38.bam
```
#  read the paths from stdin:

```
$ find  DIR -type f -name "*.bam" | ./samtools samples -m  -d 
S3	DIR/S3.bam
.	DIR/ENCFF331CGL.rnaseq.b38.bam
S2	DIR/S2.bam
S1	DIR/S1.bam
.	DIR/FAB23716.nanopore.bam
BruceWayne	DIR/retrocopy01.bwa.bam
HG02260	DIR/HG02260.transloc.chr9.14.bam
S5	DIR/S5.bam
S1	DIR/toy.bam
S4	DIR/S4.bam
```


## test if bams are indexed.

```
$ ./samtools samples -H DIR/*.samDIR/*.bam -m -i 2> /dev/null 
#SM	PATH	INDEX
.	DIR/toy.sam	0
.	DIR/ENCFF331CGL.rnaseq.b38.bam	1
.	DIR/FAB23716.nanopore.bam	1
HG02260	DIR/HG02260.transloc.chr9.14.bam	1
S1	DIR/S1.bam	1
S2	DIR/S2.bam	1
S3	DIR/S3.bam	1
S4	DIR/S4.bam	1
S5	DIR/S5.bam	1
BruceWayne	DIR/retrocopy01.bwa.bam	1
S1	DIR/toy.bam	1
```

##  find the reference

```
$ ./samtools samples DIR/*.bam -m  -d -f DIR/rotavirus_rf.fa 
.	DIR/ENCFF331CGL.rnaseq.b38.bam	.
.	DIR/FAB23716.nanopore.bam	.
HG02260	DIR/HG02260.transloc.chr9.14.bam	.
S1	DIR/S1.bam	DIR/rotavirus_rf.fa
S2	DIR/S2.bam	DIR/rotavirus_rf.fa
S3	DIR/S3.bam	DIR/rotavirus_rf.fa
S4	DIR/S4.bam	DIR/rotavirus_rf.fa
S5	DIR/S5.bam	DIR/rotavirus_rf.fa
BruceWayne	DIR/retrocopy01.bwa.bam	.
S1	DIR/toy.bam	.
```

```
$ cat refs.list  
DIR/rotavirus_rf.fa
DIR/toy.fa

$ ./samtools samples DIR/*.bam -m  -d -F  refs.list  -H
#SM	PATH	REFERENCE
.	DIR/ENCFF331CGL.rnaseq.b38.bam	.
.	DIR/FAB23716.nanopore.bam	.
HG02260	DIR/HG02260.transloc.chr9.14.bam	.
S1	DIR/S1.bam	DIR/rotavirus_rf.fa
S2	DIR/S2.bam	DIR/rotavirus_rf.fa
S3	DIR/S3.bam	DIR/rotavirus_rf.fa
S4	DIR/S4.bam	DIR/rotavirus_rf.fa
S5	DIR/S5.bam	DIR/rotavirus_rf.fa
BruceWayne	DIR/retrocopy01.bwa.bam	.
S1	DIR/toy.bam	DIR/toy.fa
```

daviesrob commented:
Looks interesting.  To get more of the checks to work, please could you rebase this onto the `develop` branch (maybe squashing it at the same time), and force-push the result back on to your `pl_sam_samples` branch in github?

You'll also need to add a line to `bam_samples.c` saying:
```
     Copyright (C) 2021 Pierre Lindenbaum
```

as we check for that along with the MIT license text in one of our tests.  You should of course substitute whoever else should own the original copyright of this file if it's not you.

lindenb commented:
@d-smirnov thanks for reviewing. I rebased my commits and merged into develop .

valeriuo commented on bam_samples.c:
```diff
@@ -0,0 +1,369 @@
+                print_error("samples", "out of memory.");
+                goto end_print;
+            }
+            kh_put(sm, sample_set, sample, &ret);
```
Check `ret` for errors.

valeriuo commented on bam_samples.c:
```diff
@@ -0,0 +1,369 @@
+    for (k = kh_begin(sample_set); k != kh_end(sample_set); ++k) {
+        if (kh_exist(sample_set, k)) {
+            char* sample = (char*)kh_key(sample_set, k);
+            if (kh_exist(sample_set, k)) free(sample);
```
The same check has been performed above.

valeriuo commented on bam_samples.c:
```diff
@@ -0,0 +1,369 @@
+        has_index = bam_idx!=NULL;
+        if (bam_idx != NULL) hts_idx_destroy(bam_idx);
+    }
+
```
Why does the method continue if the user is only interested in the presence of the index file?

lindenb replied:
well the main goal of the tool is to print `SAMPLE BAM-PATH REF?` so the index flag is just an extra column and this is why we continue. I added a comment in the code.

valeriuo commented on bam_samples.c:
```diff
@@ -0,0 +1,369 @@
+
+    /* try to load index if required */
+    if (params->test_index) {
+        hts_idx_t *bam_idx =  sam_index_load(in, fname);
```
The code should accommodate index file names that are not prefixed by the alignment file name. There could be `my_index.bam.bai` as an index for `alignments.bam`, instead of `alignments.bam.bai`. Perhaps the `-i` option should take an argument.

jmarshall commented on bam_samples.c:
```diff
@@ -0,0 +1,369 @@
+    kstring_t  *ks = NULL;
+    in = hts_open(filename,"r");
+
+    ks = &in->line;
```
Nice catch. Even better: don’t access the samFile’s private members, just declare your own kstring (not pointer) above:

```c
kstring_t ks = KS_INITIALIZE;
```

valeriuo replied:
Indeed. I thought about this, but I didn't want to be too strict, as there a few cases in bcftools of `hts_getline` using the internal kstring of the file struct. Perhaps those should be mended as well.

jmarshall replied:
Alas, bcftools is wrong to do that (though at the time there wasn't explicit documentation in _htslib/hts.h_ saying not to — there is now). There's no justification for assuming that the `htsFile` won't be using its kstring for its own purposes.

lindenb commented:
@valeriuo  thank you for the review. I had the time to have a look at comments today. I think I fixed all the comments. I added an option `-X` to specify a custom index.

```
$ ./samtools samples -X -i /home/lindenb/src/jvarkit-git/src/test/resources/S1.bam /home/lindenb/src/jvarkit-git/src/test/resources/S2.bam /home/lindenb/src/jvarkit-git/src/test/resources/S1.bam.bai /home/lindenb/src/jvarkit-git/src/test/resources/S2.bam.bai

S1	/home/lindenb/src/jvarkit-git/src/test/resources/S1.bam	1
S2	/home/lindenb/src/jvarkit-git/src/test/resources/S2.bam	1
```

when reading from `stdin` and option `-X`, we lines are expected to be in the following format: `f1.bam(tab)f1.bai`

```
$ find /home/lindenb/src/jvarkit-git/src/test/resources/ -name "S*.bam" -o -name "S*.bai" |\
    sort | paste - - | ./samtools samples -X -m -i 2> /dev/null
S1	/home/lindenb/src/jvarkit-git/src/test/resources/S1.bam	1
S2	/home/lindenb/src/jvarkit-git/src/test/resources/S2.bam	1
S3	/home/lindenb/src/jvarkit-git/src/test/resources/S3.bam	1
S4	/home/lindenb/src/jvarkit-git/src/test/resources/S4.bam	1
S5	/home/lindenb/src/jvarkit-git/src/test/resources/S5.bam	1
```

valeriuo commented on bam_samples.c:
```diff
@@ -0,0 +1,459 @@
+        }
+        ks_free(&sm_val);
+    }
+    if (count_samples == 0) {
```
The default behaviour and output should be as generic and intuitive as possible, which IMHO is: print all the samples from a file or print nothing if there isn't any. If there is an actual need for a dot (**.**) as a replacement for a sample name, the user can inspect the manual and use the extra option to enable it.

valeriuo commented on bam_samples.c:
```diff
@@ -0,0 +1,459 @@
+            goto end_print;
+        }
+    } else if (count_samples > 1 && !params->enable_multiple) {
+        print_error("samples", "Multiple @RG:\"%s\" in \"%s\". Use option -d to enable multiple", params->tag, fname);
```
Also, what is the reason for not enabling multiple sample output by default?

valeriuo commented on bam_samples.c:
```diff
@@ -0,0 +1,459 @@
+    params.test_index =0;
+
+    int opt;
+    while ((opt = getopt(argc, argv,  "hHdmiXo:f:F:t:T:")) != -1) {
```
What was `t` supposed to do?

valeriuo commented on doc/samtools-samples.1:
```diff
@@ -0,0 +1,95 @@
+'\" t
```
I've added a man page stub with the help menu taken from the code, but you should enrich it with more detailed descriptions of the command and options. Also, it would be beneficial to have the use cases from the PR notes imported here.

valeriuo commented:
Also, please make sure the tests pass when updating the PR.

jmarshall commented on bam_samples.c:
```diff
@@ -0,0 +1,459 @@
+            "       find dir1 dir2 -type f \\(-name \"*.bam\" -o -name \"*.bai\" \\) | sort | paste - - | samtools samples -X [options]\n"
+            "\n"
+            "Options:\n"
+            "  -h              print help and exit\n"
```
Better yet, use the `isatty(STDIN_FILENO)` trick to have `samtools samples` on the command line without arguments show the usage, as other samtools subcommands do.

jmarshall commented on bam_samples.c:
```diff
@@ -0,0 +1,459 @@
+            "\n"
+            "Options:\n"
+            "  -h              print help and exit\n"
+            "  -H              print a header\n"
```
Currently `view` uses `-h` for “print a header” and `cat`, `merge`, and `split` use it for something somewhat similar. Meanwhile `coverage` uses `-H` for “no header” and `view` uses `-H` for “print **only** the header”.

So if you do keep this option and you free up `-h` from being used for help, I suggest using `-h` for “print a header” as being more consistent with other samtools options than using `-H` would be.

lindenb commented:
@valeriuo @jmarshall   Thank you for your reviews, I had the time to look at the code today.

  + changed option '-h' to '-?'  to print the help
  + changed option '-H' to '-h' to add the columns header.
  + removed unused option '-t'
  + changed the description of option '-h' to print the columns header.
  + added   `if (isatty(STDIN_FILENO)) ... print usage`
  + removed `enable_multiple`, `enable_missing`
  + updated the man file.

valeriuo commented:
@lindenb any feedback on the latest changes/questions? We are trying to get this PR ready for the next release.

lindenb commented:
@valeriuo I'm currently in holidays but I'm ok with your changes ! :-)
thank you very much.
P

valeriuo commented:
Merged as 941921a41452a359e23be4678925307935961fe3 to a3914a744f3ecec199d3773dc3c730155be94ee2
